Adjust isEnabled method

kevingroeger opened the pull request:
[Issue](https://github.com/silverstripe/silverstripe-framework/issues/10799)

I added a way to be able to use a config yaml file to be able to disable certain, or all, tasks with a simple config. 

There also is a new config to set the default behaviour. This has been added to the dev.yml
```yml
SilverStripe\Dev\BuildTask:
    default_enabled: true
```

To use these changes, you can create your own yml config, example:
```yml
SilverStripe\Dev\BuildTask:
  default_enabled: false
DevTasks\ExampleTask:
  enabled: true
```

This would disable all tasks, besides the ExampleTask.

I also added the suggestion from [this](https://github.com/silverstripe/silverstripe-framework/issues/10799#issuecomment-1579554726) comment to add a deprecation notice.

Let me know if something needs to be adjusted.

Commit:
Adjust isEnabled method to use config values if present, adjust dev.yml to have a default_enabled key, add is_enabled as a private property for deprecation in cms 6

emteknetnz commented on _config/dev.yml:
```diff
@@ -1,6 +1,9 @@
 Name: DevelopmentAdmin
 ---
+SilverStripe\Dev\BuildTask:
+  default_enabled: true
```
Do we really need this `default_enabled` config?  Setting this to `false` would switch off all build tasks - that's outside the parameters of the original issue.  Seems like we should just omit this all together?

kevingroeger replied:
Yeah I wasn't too sure either, if that is *needed*, but it just turned out quite convenient when I tried it in a project with a lot of tasks. So I figured I'd add just add it. 

It's easier to put your default to false and then setting two or three tasks to enabled, instead of having to disable a lot of tasks. 

I will just omit this for now, can still readd this or something similar in the future, if needed

kevingroeger replied:
Just adding this here real quick, since you did not mark it; 

I believe we still need a fallback value if the config is not set *and* the task does not have the property $enabled set. 

I got some Tasks from addons that do not have it. Just came up when testing the changes

```php
    /**
     * @return bool
     */
    public function isEnabled()
    {
        $isEnabled = $this->config()->get('is_enabled');

        if ($isEnabled === null) {
            return static::$enabled ?? true;
        }
        return $isEnabled;
    }
```

I figured that `true` would be the smarter value as a fallback, lmk what you think

emteknetnz replied:
We don't need this `static::$enabled ?? true;` - in CMS 5 `static::enabled` is still always there are. When we do CMS 6 we'll be paying attention to this method and will refactor it as needed to allows for `static_enabled` not being there

The core idea is that all BuildTasks use `static::$enabled` right now. Very few will support the new config value `$is_enabled`. `$is_enabled` has precedence over `static::$enabled`, so `static::$enabled` is the fallback.  No need for `?? true`

emteknetnz commented on src/Dev/BuildTask.php:
```diff
@@ -32,9 +33,16 @@ public function __construct()
     private static $segment = null;
 
+    /**
+     * change this to `bool` in CMS6
```
```suggestion
     * Make this non-nullable and change this to `bool` in CMS6 with a value of `true`
```

emteknetnz commented on src/Dev/BuildTask.php:
```diff
@@ -32,9 +33,16 @@ public function __construct()
     /**
      * @var bool $enabled If set to FALSE, keep it from showing in the list
      * and from being executable through URL or CLI.
+     * @deprecated - remove in CMS6
```
```suggestion
     * @deprecated - remove in CMS 6 and rely on $is_enabled instead
```

emteknetnz commented on src/Dev/BuildTask.php:
```diff
@@ -62,9 +70,17 @@ abstract public function run($request);
      * @return bool
      */
-    public function isEnabled()
+    public function isEnabled(): bool
```
```suggestion
    public function isEnabled()
```
We can't change method signatures unless we're doing a new major version i.e. CMS 6

emteknetnz commented on src/Dev/BuildTask.php:
```diff
@@ -62,9 +70,17 @@ abstract public function run($request);
+        $isDefaultEnabled = Config::inst()->get(__CLASS__, 'default_enabled');
+
+        if ($isEnabled === null) {
+            Deprecation::notice('6.0', 'Null values not allowed in CMS6 ...');
```
```suggestion
```
In hindsight we don't actually need the the Deprecation notice, because it 99% of cases it'll be the default `null` value, and we don't want to be telling people at upgrade time that they need to set `is_enabled = true` on all of their BuildTasks since it'll become the default in CMS 6

emteknetnz commented on src/Dev/BuildTask.php:
```diff
@@ -62,9 +70,17 @@ abstract public function run($request);
+    public function isEnabled(): bool
     {
-        return $this->enabled;
+        $isEnabled = $this->config()->get('enabled');
```
```suggestion
        $isEnabled = $this->config()->get('is_enabled');
```
?

kevingroeger replied:
Ugh :D

emteknetnz commented on src/Dev/BuildTask.php:
```diff
@@ -62,9 +70,17 @@ abstract public function run($request);
     {
-        return $this->enabled;
+        $isEnabled = $this->config()->get('enabled');
+        $isDefaultEnabled = Config::inst()->get(__CLASS__, 'default_enabled');
```
As mentioned above, we probably do not want this

kevingroeger replied:
See comment above.

Commit:
remove default_enabled flag again, adjust comments as suggested, remove method signature from isEnabled, fix typo, adjust logic and add fallback value for isEnabled as true

kevingroeger commented:
Thanks a lot for reviewing! 

First time contributing to such a large project, sorry for the smaller mistakes I made ($is_enabled vs. $enabled, woops :D , unclear comments) I'm learning :) 

- Removed the default_enabled flag
- adjusted the config('$enabled') to $is_enabled
- adjusted the comments as suggested
- removed the deprecation notice
- added a fallback value (to `true`) for tasks that do not have the $enabled property set
- removed the method signature changes (force of habit 😇 )

emteknetnz commented on src/Dev/BuildTask.php:
```diff
@@ -64,7 +72,12 @@ abstract public function run($request);
+        $isEnabled = $this->config()->get('is_enabled');
+
+        if ($isEnabled === null) {
+            return static::$enabled ?? true;
```
```suggestion
            return static::$enabled;
```
As mentioned in another comment we don't need the `?? true` since `static::$enabled` is the fallback

kevingroeger replied:
I understand what you mean, but it does actually throw errors if I do not add a fallback here. 

Specifically, this error message pops up:

```
[Emergency] Uncaught Error: Access to undeclared static property DevTasks\ExampleTask::$enabled
GET /dev/tasks/
```

This happens with the tasks I wrote myself as well as with third-party tasks from addons. 

I might misunderstand something here, tbh. But I do believe we have to let this in.

emteknetnz replied:
Oh sorry my bad .. don't use `static::$enabled`, just use `$this->enabled;`

`$enabled` isn't a static property, it's a regular property

kevingroeger replied:
Oh yes, that makes an awful lot of sense :D 

Thanks for helping resolve this, I shouldn’t have missed that. My bad!

Commit:
adjust isEnabled method, change static:: to ->enabled as the property is not actually static

kevingroeger commented:
Adjusted the isEnabled method to not use static::$enabled, and instead use $this->enabled

This should be all :) 

Thanks a ton again for the massive help and the reviews. Really appreciate it!

emteknetnz reviewed (approved):
Tested locally, works well. I'll do separate PR's myself to add unit-tests and update docs/changelog

This will be released as part of the 5.1.0 release

Thanks for your contribution!
